### PR TITLE
Task 11565 payu payment feature

### PR DESCRIPTION
**Pull request by jakubzimnol**

### CODE REVIEW

File: backend/app/marshallers.py [hunk 1]
```diff
@@ -22,3 +24,30 @@
...
+
+
+order_marshaller = {
+    'id': fields.Integer,
+    'ordered_items': OrderItemsMarshaller,
+    'description': fields.String,
+    'payu_order_id': fields.String,
+    'status': fields.String,
+    'payment_status': fields.String
+}
```
Comment (h4stoor) [inline backend/app/marshallers.py:52]: You shouldn't mix classes with dicts.

File: backend/app/services.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+from app.exceptions import PayuException, BadContentInResponse, NoPermissionException, IntegrityException
+from app.models import Order, PayuStatus, Status, OrderedItem
+from app.repositories import Repository
+from init_app import db
+
+
+class PayuSender:
+    @classmethod
+    def get_access_token(cls):
```
Comment (wojciech-wolny) [inline backend/app/services.py]: don't you think this method is quite big?

File: backend/app/marshallers.py [hunk 3]
```diff
@@ -22,3 +24,30 @@
...
+    'id': fields.Integer,
+    'item': OrderItemMarshaller,
+    'quantity': fields.Integer
+}
+
+
+class OrderItemsMarshaller(fields.Raw):
+    def format(self, ordered_items):
+        return [marshal(order_item, order_item_marshaller) for order_item in ordered_items]
+
```
Comment (h4stoor) [inline backend/app/marshallers.py]: Names `OrderItemMarshaller` and `OrderItemsMarshaller` are too similar and can cause confusion.

File: backend/app/exceptions.py [hunk 4]
```diff
@@ -2,6 +2,10 @@ class ApiBaseException(Exception):
     message = None
     status_code = 400
 
+    def __init__(self, message=None):
+        if message:
+            self.message = message
+
```
Comment (h4stoor) [inline backend/app/exceptions.py:8]: You are validating `message` existence in `__init_subclass__` so this method is not necessary.

File: backend/app/services.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+        json = response.json()
+        cls.set_payu_order_id(order, json['orderId'])
+        return json['redirectUri']
+
+
+class NotificationReceiver:
+    @staticmethod
+    def get_signature(open_payu_header):
+        result1 = open_payu_header.split('signature=')
+        result2 = result1[1].split(';')
```
Comment (wojciech-wolny) [inline backend/app/services.py]: why you need to put it into a new variable, you can just overwrite result1 then you can have only result/response.

Comment (wojciech-wolny) [reply backend/app/services.py]: like 
`a = 0`
`a = a + a`

File: backend/app/services.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        if not open_payu_header:
+            raise NoPermissionException
+        signature = cls.get_signature(open_payu_header)
+        joined_str = json.dumps(json_dict) + os.environ.get('MD5')
+        expected_signature = hashlib.md5(joined_str.encode("utf-8")).hexdigest()
+        if expected_signature != signature:
+            raise NoPermissionException
+
+    @staticmethod
+    def do_nothing(order):
```
Comment (wojciech-wolny) [inline backend/app/services.py:117]: What?

**Review by wojciech-wolny — APPROVED** [review]

**Review by Rafal-Maszlej — CHANGES_REQUESTED** [review]

File: backend/app/exceptions.py (hunk 4 above)
Comment (jakubzimnol) [reply backend/app/exceptions.py:8]: This method is for change message during initialization. Sometimes I want to write another exception message which comes from request. Do you think this is bad idea?

**Review by Rafal-Maszlej — APPROVED** [review]

**Pull request merged**